rfd/0065-distributed-tracing.md

```diff
@@ -0,0 +1,247 @@
+	// main cluster, in case if validation is successful, main cluster
+	// adds remote cluster
+	ValidateTrustedCluster(context.Context, *ValidateTrustedClusterRequest) (*ValidateTrustedClusterResponse, error)
+
+	// GetDomainName returns auth server cluster name
+	GetDomainName() (string, error)
+
+	// GetClusterCACert returns the PEM-encoded TLS certs for the local cluster.
+	// If the cluster has multiple TLS certs, they will all be concatenated.
+	GetClusterCACert() (*LocalCAResponse, error)
+}
+```
+
+One of the biggest obstacles to getting **Spans** to correlate to the appropriate **Trace** will be plumbing the
+`context.Context` throughout the entire Teleport codebase. Due to the number of files that will be impacted it is
+probably best to address a few functions within an interface at a time and slowly move towards full propagation of
+`context.Context`. While this is _necessary_, it is not a **blocker** to adding tracing capabilities to Teleport - it
+only impacts the ability to correctly populate the trace tree.
+
+#### Auto-instrumentation
+
+While it is not possible to truly get automatic instrumentation in Go, we can get a decent amount of tracing for free
+by leveraging the existing [OpenTelemetry Go](https://github.com/open-telemetry/opentelemetry-go-contrib) libraries.
+We will specifically be using the `otelhttp` and `otelgrpc` libraries which automatically create spans that track inbound
+and outbound requests for `net/http` and `grpc`. These were chosen because they will provide us with proper Span Context
+propagation across the various Teleport services.
+
+In the future we may want to use other automatic instrumentation libraries like `otelaws` or `otelsql` to get better
+observability into certain parts of the code. Adding them should be on an as needed basis to help gain observability.
+
+#### Manual Instrumentation
+
+In order to instrument packages which don't already have a library available (e.g. golang.org/x/crypto/ssh) we will
+have to manually create them. All Teleport application code will also have to be manually instrumented. In addition to
+adding instrumentation to service boundaries, we will first trace the code paths along `tsh ssh` and `tsh ls` as these
+are two areas which have had the most scrutiny about latency by customers. From there further instrumentation can be
+added to other areas of the codebase, using the code already instrumented as a guide for best practices.
+
+All spans created to instrument Teleport shall follow the guidelines laid out in the
+[spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#span).
+Likewise, all attributes shall follow the [guidelines](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/common/common.md#attribute)
+to provide consistency within Teleport and across other services.
+
+#### Exporting Spans
+
+All spans created by Teleport will be exported in the
```

Comment:
From what I see Open Telemetry implements client-server model where the client (Teleport in our case) is responsible for exporting the metrics. From what I know this approach has one major problem. If the system is under heavy load then metrics tracking can significantly slow the system. From what I remember this is one of the main reason why Prometheus for example decided to just expose an endpoint and let the server to scrape the metrics whatever needed. 

Do you know how this problem is solved (or maybe still exist) in Open Telemetry? Do we have any metrics that would show how Teleport would behave in such a scenario?

Reply:
You are correct spans would be collected and exported from Teleport to an OpenTelemetry Collector. While I haven't done any testing under heavy load yet, this situation could potentially still cause exporting spans to be delayed. However, as stated in this RFD, the number of spans being sampled and exported will be determined by `sampling_rate_per_hundred`. Setting a small value is something that should be encouraged in almost all scenarios. 

I think that the sampling rate in conjunction with asynchronously exporting spans with a [BatchSpanProcessor](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/trace#NewBatchSpanProcessor) should allow us to be able to continue to export spans even under load.